packages/next/src/server/next-server.ts

```diff
@@ -1235,6 +1340,172 @@ export default class NextNodeServer extends BaseServer {
+
+            const invokeHeaders: typeof req.headers = {
+              'cache-control': '',
+              ...req.headers,
+              'x-invoke-path': invokePathname,
+              'x-invoke-query': encodeURIComponent(invokeQuery),
+            }
+
+            const forbiddenHeaders = (global as any).__NEXT_USE_UNDICI
+              ? [
+                  'content-length',
+                  'keepalive',
+                  'content-encoding',
+                  'transfer-encoding',
+                  // https://github.com/nodejs/undici/issues/1470
+                  'connection',
+                ]
+              : [
+                  'content-length',
+                  'keepalive',
+                  'content-encoding',
+                  'transfer-encoding',
+                ]
+
+            for (const key of forbiddenHeaders) {
+              delete invokeHeaders[key]
+            }
+
+            const invokeRes = await fetch(renderUrl, {
+              method: req.method,
+              headers: invokeHeaders as any,
+              redirect: 'manual',
+              ...(req.method !== 'GET' && req.method !== 'HEAD'
+                ? {
+                    // @ts-ignore
+                    duplex: 'half',
+                    body: getRequestMeta(
+                      req,
+                      '__NEXT_CLONABLE_BODY'
+                    )?.cloneBodyStream() as any as ReadableStream,
+                  }
+                : {}),
+            })
+
+            const noFallback = invokeRes.headers.get('x-no-fallback')
+
+            if (noFallback) {
+              if (bubbleNoFallback) {
+                return { finished: false }
+              } else {
+                await this.render404(req, res, parsedUrl)
+                return {
+                  finished: true,
+                }
+              }
+            }
+
+            for (const [key, value] of Object.entries(
+              toNodeHeaders(invokeRes.headers)
+            )) {
+              if (
+                ![
+                  'content-encoding',
+                  'transfer-encoding',
+                  'keep-alive',
+                  'connection',
+                ].includes(key) &&
+                value !== undefined
+              ) {
+                if (key === 'set-cookie') {
+                  const curValue = res.getHeader(key)
+                  const newValue: string[] = [] as string[]
+                  for (const cookie of splitCookiesString(curValue || '')) {
+                    newValue.push(cookie)
+                  }
+                  for (const val of (Array.isArray(value)
+                    ? value
+                    : value
+                    ? [value]
+                    : []) as string[]) {
+                    newValue.push(val)
+                  }
+                  res.setHeader(key, newValue)
+                } else {
+                  res.setHeader(key, value as string)
+                }
```

Comment:
This beheviour is actually already handled by the `toNodeHeaders` function.

https://github.com/vercel/next.js/blob/400ccf7b1c802c94127d8d8e0d5e9bdf9aab270c/packages/next/src/server/web/utils.ts#L95-L112

Reply:
This needs to merge new headers with the existing `res` headers which the `toNodeHelpers` does not currently do, could be adapted to though.